### ISSUE TITLE
Regression Bug in MySql involving unions used inside 'in' clauses

### ISSUE THREAD

**Issue by timgautier** [slick/slick#2104]
**DESCRIPTION:**
If a `Query` utilizes a union and that `Query` is then used inside an `in` clause, MySQL will throw a syntax error. I noticed this bug when I tried to upgrade to 3.3.1 from 3.2.0. Code that had previously worked fine quit working. After some digging I found that it seems to have been introduced [here](https://github.com/slick/slick/commit/c9c6dd08a37533711ad9fd24c4c52d7f5ff66300).

**STEPS TO REPRODUCE:**
Add this unit test to UnionTest.scala to see the error. Note that it only seems to be an issue in MySQL.
```
def testInUnion = {
    class T(tag: Tag) extends Table[(String, Int)](tag, "t") {
      def id = column[String]("id")
      def data = column[Int]("data")
      def * = (id, data)
    }

    val ts = TableQuery[T]
    val q1 = ts.filter(_.data.in((ts.filter(_.id === "a") ++ ts.filter(_.id === "c")).map(_.data)))
    DBIO.seq(
      ts.schema.create,
      ts.map(f => (f.id, f.data)) ++= Seq(("a", 1), ("b", 2), ("c", 1), ("d", 1)),
      q1.result.map(_ shouldBe Vector(("a", 1), ("c", 1), ("d", 1)))
    )
  }
```

**EXPECTED RESULT:**

Passed Test

**ACTUAL RESULT:**

```
*** (s.j.J.statementAndParameter) Executing prepared statement: com.mysql.jdbc.JDBC42PreparedStatement@2a596a45: select `id`, `data`
from `t`
where `data` in (
  (
    select `data` as x2
    from `t`
    where `id` = 'a'
  )
  union all (
    select `data` as x2
    from `t`
    where `id` = 'c'
  )
)
[error] Test com.typesafe.slick.testkit.tests.UnionTest.testInUnion[mysql] failed: com.mysql.jdbc.exceptions.jdbc4.MySQLSyntaxErrorException: You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near 'union all (
[error]     select `data` as x2
[error]     from `t`
[error]     where `id` = 'c'
[error]   )
[error] )' at line 9, took 0.017 sec
[error]     at jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
[error]     at jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
[error]     at jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
[error]     at java.lang.reflect.Constructor.newInstance(Constructor.java:490)
[error]     at com.mysql.jdbc.Util.handleNewInstance(Util.java:425)
[error]     at com.mysql.jdbc.Util.getInstance(Util.java:408)
[error]     at com.mysql.jdbc.SQLError.createSQLException(SQLError.java:944)
[error]     at com.mysql.jdbc.MysqlIO.checkErrorPacket(MysqlIO.java:3976)
[error]     at com.mysql.jdbc.MysqlIO.checkErrorPacket(MysqlIO.java:3912)
[error]     at com.mysql.jdbc.MysqlIO.sendCommand(MysqlIO.java:2530)
[error]     at com.mysql.jdbc.MysqlIO.sqlQueryDirect(MysqlIO.java:2683)
[error]     at com.mysql.jdbc.ConnectionImpl.execSQL(ConnectionImpl.java:2486)
[error]     at com.mysql.jdbc.PreparedStatement.executeInternal(PreparedStatement.java:1858)
[error]     at com.mysql.jdbc.PreparedStatement.execute(PreparedStatement.java:1197)
[error]     at slick.jdbc.LoggingPreparedStatement$$anonfun$execute$5.apply$mcZ$sp(LoggingStatement.scala:160)
[error]     at slick.jdbc.LoggingPreparedStatement$$anonfun$execute$5.apply(LoggingStatement.scala:160)
[error]     at slick.jdbc.LoggingPreparedStatement$$anonfun$execute$5.apply(LoggingStatement.scala:160)
[error]     at slick.jdbc.LoggingStatement.logged(LoggingStatement.scala:87)
[error]     at slick.jdbc.LoggingPreparedStatement.execute(LoggingStatement.scala:160)
[error]     at slick.jdbc.StatementInvoker.results(StatementInvoker.scala:39)
[error]     at slick.jdbc.StatementInvoker.iteratorTo(StatementInvoker.scala:22)
[error]     at slick.jdbc.Invoker$class.foreach(Invoker.scala:47)
[error]     at slick.jdbc.StatementInvoker.foreach(StatementInvoker.scala:16)
[error]     at slick.jdbc.StreamingInvokerAction$class.run(StreamingInvokerAction.scala:22)
[error]     at slick.jdbc.JdbcActionComponent$QueryActionExtensionMethodsImpl$$anon$1.run(JdbcActionComponent.scala:214)
[error]     at slick.jdbc.JdbcActionComponent$QueryActionExtensionMethodsImpl$$anon$1.run(JdbcActionComponent.scala:214)
[error]     at slick.basic.BasicBackend$DatabaseDef$$anon$2.liftedTree1$1(BasicBackend.scala:276)
[error]     at slick.basic.BasicBackend$DatabaseDef$$anon$2.run(BasicBackend.scala:276)
[error]     at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
[error]     at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
[error]     at java.lang.Thread.run(Thread.java:834)
```

**SETUP:**

This was running on MariaDB 10.3.13, but this query will throw an error in MySQL also.

Ultimately this is caused by MySQL's Bison parser getting confused by the parentheses.
This query is valid in most databases, but the first set of parentheses of the `union all` causes a syntax error in MySQL.
```
select `id`, `data`
from `t`
where `data` in (
  (
    select `data` as x2
    from `t`
    where `id` = 'a'
  )
  union all (
    select `data` as x2
    from `t`
    where `id` = 'c'
  )
)
```

If we take those parentheses out, we have a working query.
```
select `id`, `data`
from `t`
where `data` in (
    select `data` as x2
    from `t`
    where `id` = 'a'
  union all (
    select `data` as x2
    from `t`
    where `id` = 'c'
  )
)
```
The 2nd set of parentheses is irrelevant.

An easy solution to this issue is to change the `Union` case in the `expr` function of `MySQLProfile.scala` like so:
```
     case Union(left, right, all) =>
        b"\{"
        buildFrom(left, None, true)
        if (all) b"\nunion all " else b"\nunion "
        buildFrom(right, None)
        b"\}"
```
However, this will partially reintroduce the bug Anandhan fixed with his change and cause `UnionAdditionalTest.testOrderByWithUnion` and `UnionAdditionalTest.testLimitWithUnion` to fail. (I say 'partially', because we can keep the parentheses around the second of the subqueries, but not the first.)

In fact, I'm not sure how to get all three of these tests to pass. If two queries with `limit` or `order by` are `union`ed together, they have to be surrounded by parentheses or you'll get a syntax error. However, if you take that syntactically correct query and use it in an `in` clause, it can't have parentheses around the first of the two subqueries, or you'll get a syntax error. It seems to be a catch 22.

As an example of this catch 22, try adding this unit test to the end of `UnionAdditionalTest.scala`.
```
  def testLimitUnionWithinInClause = {
    val q1 = for (m <- managers drop 1 take 2) yield m.id
    val q2 = for (e <- employees drop 1 take 3) yield e.id
    val q3 = managers.filter(_.id.in(q1.union(q2))).map(_.id)
    (for {
      _ <- (managers.schema ++ employees.schema).create
      _ <- managers ++= Seq((1, "Peter", "HR"), (2, "Amy", "IT"), (3, "Steve", "IT"))
      _ <- employees ++= Seq((4, "Leonard", 2), (5, "Jennifer", 1), (6, "Tom", 1), (7, "Ben", 1), (8, "Greg", 3))
      _ <- mark("q1", q1.result).map(r => r.toSet shouldBe Set(2,3))
      _ <- mark("q2", q2.result).map(r => r.toSet shouldBe Set(5,6,7))
      _ <- mark("q3", q3.result).map(_ shouldBe Vector(2,3))
    } yield ()) andFinally (managers.schema ++ employees.schema).drop
  }
```
I'm not sure this test *can* pass in MySQL, though it seems to pass fine in other databases.

Bottom line, I'm not sure where to go with this. We use slick primarily for its ability to compose queries, which we use abundantly. This seems to be a case where the syntactic validity of an inner query is dependent on how it's used in an outer query, which seems to make composition a bit of a run time landmine. Thoughts?

**Comment by timgautier** [comment]
I submitted a PR that works around the issue by wrapping the union subquery inside the 'in' clause with another select. This seems to be the only way to maintain composability.

As an example, this is syntactically valid in MySQL:
```
(select data
   from some_table
   limit 1)
union
(select data
   from another_table
   limit 1);
```

However, putting that same query inside an 'in' clause will cause that query to throw a syntax error:
```
select *
from a_third_table
where data in (
  (select data
     from some_table
     limit 1)
  union
  (select data
     from another_table
     limit 1)
);
```
Slick will allow you to compose these queries together, but MySQL will not. The only real solution I can find is to detect this situation and wrap the subquery like so:
```
select *
from a_third_table
where data in (
  select t.* from (
    (select data
       from some_table
       limit 1)
    union
    (select data
       from another_table
       limit 1)
  ) t
);
```
This meets the intention of the query without the syntax error and restores composability.

In this case I've aliased the subquery as `t` and selected `t.*`. This could probably be accomplished in a more Slick idiomatic way, but I have limited experience with the Slick code base and couldn't find that solution. However, due to the specificity of this solution the `*` should always be a single column and the `t` should always be alone in scope with no chance of conflict.

**Comment by timgautier** [comment]
Any chance this will ever be fixed or should we fork our own version of slick?

**Comment by timgautier** [comment]
Bueller? Bueller? 

I would have figured a regression bug that's been identified, reported, fixed, and PRed would have at least earned a comment by now.

**Comment by SethTisue** [comment]
@timgautier there was a lapse in maintenance in this repo, but things are moving again.
